### DISCUSSION TITLE
Scikit-build under cibuildwheel problem

### DISCUSSION THREAD

**Discussion by Czaki** [pypa/cibuildwheel#1232] [Q&A]
@henryiii I try to make PR for migrating SimpleITK python wheel build to cibuildwheel for reducing the delay of release for new python version.

When I build locally https://github.com/SimpleITK/SimpleITKPythonPackage everything works ok. But if I run `CIBW_BUILD="cp310*" cibuildwheel --platform=linux .` build fails because `PYTHON_LIBRARY` is set to file name (`'libpython3.10.a'`), not the full path to file. 

When I read scikit-build documentation, I understand that it should contain the full path. And it looks like it has it when built locally. 

Did I wrong understand, or is there a reason why running it under cibuildwheel makes the build fail?  

Or should I move this to `scikit-build` as an issue?

**Comment by henryiii** [comment]
A quick check, does https://github.com/scikit-build/scikit-build/pull/739 fix it? (Actually, I might have forgotten to add the path itself in that case, but if it triggers a test failure, that would be helpful. So maybe try it if the scikit-build tests pass.)

**Reply by Czaki** [reply]
```
        PYTHON_LIBRARY variable is defined but corresponds to nonexistent file
        '/project/libpython3.10.a',
```
abspath uses the current directory, not the library directory.

**Reply by henryiii** [reply]
I was hoping that would break scikit-build's tests when I realized I didn't add the libdir to it. :( I can try that as long as you can test, really wish scikit-build's tests would catch this sort of thing. At least it seems to be having an effect, so this would probably be a fix.

**Reply by henryiii** [reply]
If I want to run this locally, can I just use `CIBW_BUILD="cp310*" cibuildwheel --platform=linux .` on the current repo as-is? Trying to figure out which path it follows to get the libdir.

**Reply by Czaki** [reply]
my only change is in 176 line of `CMakeLists.txt` to print current value of variables:
```
    message(FATAL_ERROR "PYTHON_LIBRARY variable is defined but corresponds to nonexistent file '${PYTHON_LIBRARY}', '${PYTHON_INCLUDE_DIR}' '${PYTHON_EXECUTABLE}")
```

But this is only for debug, so you could start from repo as is. 

as I read the code of `_guess_python_library`, it is possible that it ends without determining the proper path and returns only the name. 

https://github.com/scikit-build/scikit-build/blob/a54f4fc66ec361f27804641cac0f31dec2a4617d/skbuild/cmaker.py#L509

I could test it on your request.

**Reply by henryiii** [reply]
Yes, I was just thinking that! Are you building this in manylinux? There's no static library in manylinux, you are only supposed to build extensions in manylinux and those shouldn't be linked to libpython.

**Reply by Czaki** [reply]
I'm not the maintainer of this project, but I'm the maintainer of the project that depends on it and waiting half a year until they update their custom scripts and release a new wheel for new python. 

As they are open for introduce cibuildwheel I decide to try to do this: https://github.com/SimpleITK/SimpleITK/issues/1642

Based on your remark, I will try to better check their build systems and understand how they do it in manylinux currently https://github.com/SimpleITK/SimpleITK/tree/master/Utilities/Distribution/manylinux

I had hoped that if they have `setup.py` entry point in [SimpleITK/SimpleITKPythonPackage](https://github.com/SimpleITK/SimpleITKPythonPackage) it will be simpler.

**Reply by henryiii** [reply]
From a quick look, it's quite possibly they are accessing and passing around PYTHON_LIBRARY when they don't need to. You should not need to to build a Python extension - the symbols are already loaded in the interpreter. As a first try I'd delete every line with PYTHON_LIBRARY in it in the CMakeLists.txt.
